### PR TITLE
Add smoke tests for all the examples, fix 17 broken examples.

### LINKED ISSUE #83987: MaterialStateOutlinedBorder interactive app is broken
Hi the example of MaterialStateOutlinedBorder doesn't run in the browser.

It logs following:

```
Error: Assertion failed: file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter-sdk/packages/flutter/lib/src/rendering/mouse_tracker.dart:222:12
(event is PointerAddedEvent) == (lastEvent is PointerRemovedEvent)
is not trueError: Error: Assertion failed: file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter-sdk/packages/flutter/lib/src/rendering/mouse_tracker.dart:222:12
(event is PointerAddedEvent) == (lastEvent is PointerRemovedEvent)
is not true
```

https://api.flutter.dev/flutter/material/MaterialStateOutlinedBorder-class.html

grafik

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This adds a smoke test for every single API example. It also fixes 17 tests that had bugs in them, or were otherwise broken, and even fixes one actual bug in the framework, and one limitation in the framework.

The bug in the framework is that `NetworkImage`'s `_loadAsync` method had `await response.drain<List<int>>();`, but if the response is null, it will throw a cryptic exception saying that `Null` can't be assigned to `List<int>`. The fix was just to use `await response.drain<void>();` instead.

The limitation is that `RelativePositionedTransition` takes an `Animation<Rect> rect` parameter, and if you want to use a `RectTween` with it, the value emitted there is `Rect?`, and one of the examples was just casting from `Animation<Rect>` to `Animation<Rect?>`, which is invalid, so I modified `RelativePositionedTransition` to take a `Rect?` and just use `Rect.zero` if the `rect` is null.

**Epilepsy warning: the following image flickers a lot.**

https://user-images.githubusercontent.com/8867023/131196265-f78917c2-c979-4ebf-b928-b36c96630161.mp4

## Related Issues

 - Fixes https://github.com/flutter/flutter/issues/83987

## Tests
 - Add smoke tests for 239 example apps.

### CODE REVIEW

File: dev/tools/examples_smoke_test.dart [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+const bool kIsWeb = identical(0, 0.0);
+FileSystem filesystem = const LocalFileSystem();
+ProcessManager processManager = const LocalProcessManager();
+Platform platform = const LocalPlatform();
+
+FutureOr<dynamic> main() async {
+  if (!platform.isLinux && !platform.isWindows && !platform.isMacOS) {
+    stderr.writeln('Example smoke tests are only designed to run on desktop platforms');
+    exitCode = 4;
```
Comment (christopherfujino) [inline dev/tools/examples_smoke_test.dart:28]: TIL about `exitCode`. Is this equivalent to `exit(4);`?

File: dev/tools/examples_smoke_test.dart [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+// Generates the combined smoke test.
+Future<File> generateTest(Directory apiDir) async {
+  final Directory examplesLibDir = apiDir.childDirectory('lib');
+
+  // Get files from git, to avoid any non-repo files that might be in someone's
+  // workspace.
+  final List<String> gitFiles = (await runCommand(
+    <String>['git', 'ls-files', '**/*.dart'],
```
Comment (christopherfujino) [inline dev/tools/examples_smoke_test.dart:111]: does the glob and path separator work on windows?

File: dev/tools/examples_smoke_test.dart (hunk 1 above)
Comment (gspencergoog) [reply dev/tools/examples_smoke_test.dart:28]: Yes, except that it allows for a clean shutdown.  If you call exit(4), then if the main is wrapped (e.g. in an integration test) then the test shutdown doesn't occur cleanly (it just exits the app immediately), and not all results are reported.

File: dev/tools/examples_smoke_test.dart (hunk 2 above)
Comment (gspencergoog) [reply dev/tools/examples_smoke_test.dart:111]: Yes, in git it does.

Comment (christopherfujino) [reply dev/tools/examples_smoke_test.dart:111]: ahh, so the glob is expanded by git?

**Review by christopherfujino — APPROVED** [review]
LGTM

File: dev/tools/examples_smoke_test.dart (hunk 2 above)
Comment (gspencergoog) [reply dev/tools/examples_smoke_test.dart:111]: Yep.

File: packages/flutter/lib/src/painting/_network_image_io.dart [hunk 3]
```diff
@@ -96,7 +96,7 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
         // The network may be only temporarily unavailable, or the file will be
         // added on the server later. Avoid having future calls to resolve
         // fail to check the network again.
-        await response.drain<List<int>>();
+        await response.drain<void>();
```
Comment (Piinks) [inline packages/flutter/lib/src/painting/_network_image_io.dart:99]: Is there a test for this? Or does the sample code test cover it?

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 4]
```diff
@@ -732,11 +732,13 @@ class RelativePositionedTransition extends AnimatedWidget {
...
   /// The animation that controls the child's size and position.
   ///
+  /// If the animation returns a null [Rect], the rect is assumed to be [Rect.zero].
+  ///
   /// See also:
   ///
   ///  * [size], which gets the size of the box that the [Positioned] widget's
   ///    offsets are relative to.
-  Animation<Rect> get rect => listenable as Animation<Rect>;
+  Animation<Rect?> get rect => listenable as Animation<Rect?>;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/transitions.dart:741]: Same here, does this need a test? Or does the sample code test ensure this?

**Review by Piinks — APPROVED** [review]
LGTM after confirming test coverage, but I think since the sample code tests exposed those issues, I am guessing they provide test coverage

**Pull request merged**